File: distribution/python/dqops/client/models/table_monthly_recurring_check_categories_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

import attr

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.table_accuracy_monthly_recurring_checks_spec import (
        TableAccuracyMonthlyRecurringChecksSpec,
    )
    from ..models.table_availability_monthly_recurring_checks_spec import (
        TableAvailabilityMonthlyRecurringChecksSpec,
    )
    from ..models.table_monthly_recurring_check_categories_spec_comparisons import (
        TableMonthlyRecurringCheckCategoriesSpecComparisons,
    )
    from ..models.table_monthly_recurring_check_categories_spec_custom import (
        TableMonthlyRecurringCheckCategoriesSpecCustom,
    )
    from ..models.table_schema_monthly_recurring_checks_spec import (
        TableSchemaMonthlyRecurringChecksSpec,
    )
    from ..models.table_sql_monthly_recurring_checks_spec import (
        TableSqlMonthlyRecurringChecksSpec,
    )
    from ..models.table_timeliness_monthly_recurring_checks_spec import (
        TableTimelinessMonthlyRecurringChecksSpec,
    )
    from ..models.table_volume_monthly_recurring_checks_spec import (
        TableVolumeMonthlyRecurringChecksSpec,
    )


T = TypeVar("T", bound="TableMonthlyRecurringCheckCategoriesSpec")


@attr.s(auto_attribs=True)
class TableMonthlyRecurringCheckCategoriesSpec:
    """
    Attributes:
        custom (Union[Unset, TableMonthlyRecurringCheckCategoriesSpecCustom]): Dictionary of custom checks. The keys are
            check names.
        volume (Union[Unset, TableVolumeMonthlyRecurringChecksSpec]):
        timeliness (Union[Unset, TableTimelinessMonthlyRecurringChecksSpec]):
        accuracy (Union[Unset, TableAccuracyMonthlyRecurringChecksSpec]):
        sql (Union[Unset, TableSqlMonthlyRecurringChecksSpec]):
        availability (Union[Unset, TableAvailabilityMonthlyRecurringChecksSpec]):
        schema (Union[Unset, TableSchemaMonthlyRecurringChecksSpec]):
        comparisons (Union[Unset, TableMonthlyRecurringCheckCategoriesSpecComparisons]): Dictionary of configuration of
            checks for table comparisons. The key that identifies each comparison must match the name of a data comparison
            that is configured on the parent table.
    """

    custom: Union[Unset, "TableMonthlyRecurringCheckCategoriesSpecCustom"] = UNSET
    volume: Union[Unset, "TableVolumeMonthlyRecurringChecksSpec"] = UNSET
    timeliness: Union[Unset, "TableTimelinessMonthlyRecurringChecksSpec"] = UNSET
    accuracy: Union[Unset, "TableAccuracyMonthlyRecurringChecksSpec"] = UNSET
    sql: Union[Unset, "TableSqlMonthlyRecurringChecksSpec"] = UNSET
    availability: Union[Unset, "TableAvailabilityMonthlyRecurringChecksSpec"] = UNSET
    schema: Union[Unset, "TableSchemaMonthlyRecurringChecksSpec"] = UNSET
    comparisons: Union[
        Unset, "TableMonthlyRecurringCheckCategoriesSpecComparisons"
    ] = UNSET
    additional_properties: Dict[str, Any] = attr.ib(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        custom: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom, Unset):
            custom = self.custom.to_dict()

        volume: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.volume, Unset):
            volume = self.volume.to_dict()

        timeliness: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.timeliness, Unset):
            timeliness = self.timeliness.to_dict()

        accuracy: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.accuracy, Unset):
            accuracy = self.accuracy.to_dict()

        sql: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.sql, Unset):
            sql = self.sql.to_dict()

        availability: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.availability, Unset):
            availability = self.availability.to_dict()

        schema: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.schema, Unset):
            schema = self.schema.to_dict()

        comparisons: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.comparisons, Unset):
            comparisons = self.comparisons.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if custom is not UNSET:
            field_dict["custom"] = custom
        if volume is not UNSET:
            field_dict["volume"] = volume
        if timeliness is not UNSET:
            field_dict["timeliness"] = timeliness
        if accuracy is not UNSET:
            field_dict["accuracy"] = accuracy
        if sql is not UNSET:
            field_dict["sql"] = sql
        if availability is not UNSET:
            field_dict["availability"] = availability
        if schema is not UNSET:
            field_dict["schema"] = schema
        if comparisons is not UNSET:
            field_dict["comparisons"] = comparisons

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.table_accuracy_monthly_recurring_checks_spec import (
            TableAccuracyMonthlyRecurringChecksSpec,
        )
        from ..models.table_availability_monthly_recurring_checks_spec import (
            TableAvailabilityMonthlyRecurringChecksSpec,
        )
        from ..models.table_monthly_recurring_check_categories_spec_comparisons import (
            TableMonthlyRecurringCheckCategoriesSpecComparisons,
        )
        from ..models.table_monthly_recurring_check_categories_spec_custom import (
            TableMonthlyRecurringCheckCategoriesSpecCustom,
        )
        from ..models.table_schema_monthly_recurring_checks_spec import (
            TableSchemaMonthlyRecurringChecksSpec,
        )
        from ..models.table_sql_monthly_recurring_checks_spec import (
            TableSqlMonthlyRecurringChecksSpec,
        )
        from ..models.table_timeliness_monthly_recurring_checks_spec import (
            TableTimelinessMonthlyRecurringChecksSpec,
        )
        from ..models.table_volume_monthly_recurring_checks_spec import (
            TableVolumeMonthlyRecurringChecksSpec,
        )

        d = src_dict.copy()
        _custom = d.pop("custom", UNSET)
        custom: Union[Unset, TableMonthlyRecurringCheckCategoriesSpecCustom]
        if isinstance(_custom, Unset):
            custom = UNSET
        else:
            custom = TableMonthlyRecurringCheckCategoriesSpecCustom.from_dict(_custom)

        _volume = d.pop("volume", UNSET)
        volume: Union[Unset, TableVolumeMonthlyRecurringChecksSpec]
        if isinstance(_volume, Unset):
            volume = UNSET
        else:
            volume = TableVolumeMonthlyRecurringChecksSpec.from_dict(_volume)

        _timeliness = d.pop("timeliness", UNSET)
        timeliness: Union[Unset, TableTimelinessMonthlyRecurringChecksSpec]
        if isinstance(_timeliness, Unset):
            timeliness = UNSET
        else:
            timeliness = TableTimelinessMonthlyRecurringChecksSpec.from_dict(
                _timeliness
            )

        _accuracy = d.pop("accuracy", UNSET)
        accuracy: Union[Unset, TableAccuracyMonthlyRecurringChecksSpec]
        if isinstance(_accuracy, Unset):
            accuracy = UNSET
        else:
            accuracy = TableAccuracyMonthlyRecurringChecksSpec.from_dict(_accuracy)

        _sql = d.pop("sql", UNSET)
        sql: Union[Unset, TableSqlMonthlyRecurringChecksSpec]
        if isinstance(_sql, Unset):
            sql = UNSET
        else:
            sql = TableSqlMonthlyRecurringChecksSpec.from_dict(_sql)

        _availability = d.pop("availability", UNSET)
        availability: Union[Unset, TableAvailabilityMonthlyRecurringChecksSpec]
        if isinstance(_availability, Unset):
            availability = UNSET
        else:
            availability = TableAvailabilityMonthlyRecurringChecksSpec.from_dict(
                _availability
            )

        _schema = d.pop("schema", UNSET)
        schema: Union[Unset, TableSchemaMonthlyRecurringChecksSpec]
        if isinstance(_schema, Unset):
            schema = UNSET
        else:
            schema = TableSchemaMonthlyRecurringChecksSpec.from_dict(_schema)

        _comparisons = d.pop("comparisons", UNSET)
        comparisons: Union[Unset, TableMonthlyRecurringCheckCategoriesSpecComparisons]
        if isinstance(_comparisons, Unset):
            comparisons = UNSET
        else:
            comparisons = TableMonthlyRecurringCheckCategoriesSpecComparisons.from_dict(
                _comparisons
            )

        table_monthly_recurring_check_categories_spec = cls(
            custom=custom,
            volume=volume,
            timeliness=timeliness,
            accuracy=accuracy,
            sql=sql,
            availability=availability,
            schema=schema,
            comparisons=comparisons,
        )

        table_monthly_recurring_check_categories_spec.additional_properties = d
        return table_monthly_recurring_check_categories_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/checks/table/recurring/TableMonthlyRecurringCheckCategoriesSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.checks.table.recurring;

import com.dqops.checks.AbstractRootChecksContainerSpec;
import com.dqops.checks.CheckTarget;
import com.dqops.checks.CheckTimeScale;
import com.dqops.checks.CheckType;
import com.dqops.checks.table.recurring.accuracy.TableAccuracyMonthlyRecurringChecksSpec;
import com.dqops.checks.table.recurring.availability.TableAvailabilityMonthlyRecurringChecksSpec;
import com.dqops.checks.table.recurring.comparison.TableComparisonMonthlyRecurringChecksSpecMap;
import com.dqops.checks.table.recurring.schema.TableSchemaMonthlyRecurringChecksSpec;
import com.dqops.checks.table.recurring.sql.TableSqlMonthlyRecurringChecksSpec;
import com.dqops.checks.table.recurring.volume.TableVolumeMonthlyRecurringChecksSpec;
import com.dqops.checks.table.recurring.timeliness.TableTimelinessMonthlyRecurringChecksSpec;
import com.dqops.metadata.timeseries.TimeSeriesConfigurationProvider;
import com.dqops.metadata.timeseries.TimeSeriesConfigurationSpec;
import com.dqops.metadata.timeseries.TimePeriodGradient;
import com.dqops.metadata.timeseries.TimeSeriesMode;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.scheduling.CheckRunRecurringScheduleGroup;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.utils.serialization.IgnoreEmptyYamlSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.EqualsAndHashCode;

import java.util.Objects;

/**
 * Container of table level monthly recurring checks. Contains categories of monthly recurring checks.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class TableMonthlyRecurringCheckCategoriesSpec extends AbstractRootChecksContainerSpec implements TimeSeriesConfigurationProvider {
    public static final ChildHierarchyNodeFieldMapImpl<TableMonthlyRecurringCheckCategoriesSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractRootChecksContainerSpec.FIELDS) {
        {
            put("volume", o -> o.volume);
            put("timeliness", o -> o.timeliness);
            put("accuracy", o -> o.accuracy);
            put("sql", o -> o.sql);
            put("availability", o -> o.availability);
            put("schema", o -> o.schema);
            put("comparisons", o -> o.comparisons);
        }
    };

    @JsonPropertyDescription("Monthly recurring of volume data quality checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableVolumeMonthlyRecurringChecksSpec volume;

    @JsonPropertyDescription("Monthly recurring of timeliness checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableTimelinessMonthlyRecurringChecksSpec timeliness;

    @JsonPropertyDescription("Monthly recurring accuracy checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableAccuracyMonthlyRecurringChecksSpec accuracy;

    @JsonPropertyDescription("Monthly recurring of custom SQL checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableSqlMonthlyRecurringChecksSpec sql;

    @JsonPropertyDescription("Daily partitioned availability checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableAvailabilityMonthlyRecurringChecksSpec availability;

    @JsonPropertyDescription("Monthly recurring table schema checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableSchemaMonthlyRecurringChecksSpec schema;

    @JsonPropertyDescription("Dictionary of configuration of checks for table comparisons. The key that identifies each comparison must match the name of a data comparison that is configured on the parent table.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableComparisonMonthlyRecurringChecksSpecMap comparisons = new TableComparisonMonthlyRecurringChecksSpecMap();

    /**
     * Returns the container of recurring for volume data quality checks.
     * @return Container of row volume data quality recurring.
     */
    public TableVolumeMonthlyRecurringChecksSpec getVolume() {
        return volume;
    }

    /**
     * Sets the container of volume data quality checks (recurring).
     * @param volume New volume checks.
     */
    public void setVolume(TableVolumeMonthlyRecurringChecksSpec volume) {
        this.setDirtyIf(!Objects.equals(this.volume, volume));
        this.volume = volume;
        this.propagateHierarchyIdToField(volume, "volume");
    }

    /**
     * Returns a container of table level timeliness recurring.
     * @return Custom timeliness recurring.
     */
    public TableTimelinessMonthlyRecurringChecksSpec getTimeliness() {
        return timeliness;
    }

    /**
     * Sets a reference to a container of timeliness recurring.
     * @param timeliness Custom timeliness recurring.
     */
    public void setTimeliness(TableTimelinessMonthlyRecurringChecksSpec timeliness) {
        this.setDirtyIf(!Objects.equals(this.timeliness, timeliness));
        this.timeliness = timeliness;
        this.propagateHierarchyIdToField(timeliness, "timeliness");
    }

    /**
     * Returns a container of table level accuracy recurring checks.
     * @return Table level accuracy checks.
     */
    public TableAccuracyMonthlyRecurringChecksSpec getAccuracy() {
        return accuracy;
    }

    /**
     * Sets a new container of monthly recurring accuracy checks.
     * @param accuracy New daily recurring accuracy checks.
     */
    public void setAccuracy(TableAccuracyMonthlyRecurringChecksSpec accuracy) {
        this.setDirtyIf(!Objects.equals(this.accuracy, accuracy));
        this.accuracy = accuracy;
        this.propagateHierarchyIdToField(accuracy, "accuracy");
    }

    /**
     * Returns a container of custom sql recurring.
     * @return Container of custom sql recurring.
     */
    public TableSqlMonthlyRecurringChecksSpec getSql() {
        return sql;
    }

    /**
     * Sets a reference to a container of custom sql recurring.
     * @param sql Custom sql recurring.
     */
    public void setSql(TableSqlMonthlyRecurringChecksSpec sql) {
        this.setDirtyIf(!Objects.equals(this.sql, sql));
        this.sql = sql;
        this.propagateHierarchyIdToField(sql, "sql");
    }

    /**
     * Returns a container of custom sql checks.
     * @return Custom sql checks.
     */
    public TableAvailabilityMonthlyRecurringChecksSpec getAvailability() {
        return availability;
    }

    /**
     * Sets a reference to a container of custom sql checks.
     * @param availability Container of custom sql checks.
     */
    public void setAvailability(TableAvailabilityMonthlyRecurringChecksSpec availability) {
        this.setDirtyIf(!Objects.equals(this.availability, availability));
        this.availability = availability;
        this.propagateHierarchyIdToField(availability, "availability");
    }

    /**
     * Returns a container of table schema checks.
     * @return Table schema checks.
     */
    public TableSchemaMonthlyRecurringChecksSpec getSchema() {
        return schema;
    }

    /**
     * Sets a reference to a container with the table schema checks.
     * @param schema Container of table schema checks.
     */
    public void setSchema(TableSchemaMonthlyRecurringChecksSpec schema) {
        this.setDirtyIf(!Objects.equals(this.schema, schema));
        this.schema = schema;
        this.propagateHierarchyIdToField(schema, "schema");
    }

    /**
     * Returns the dictionary of comparisons.
     * @return Dictionary of comparisons.
     */
    @Override
    public TableComparisonMonthlyRecurringChecksSpecMap getComparisons() {
        return comparisons;
    }

    /**
     * Sets the dictionary of comparisons.
     * @param comparisons Dictionary of comparisons.
     */
    public void setComparisons(TableComparisonMonthlyRecurringChecksSpecMap comparisons) {
        this.setDirtyIf(!Objects.equals(this.comparisons, comparisons));
        this.comparisons = comparisons;
        this.propagateHierarchyIdToField(comparisons, "comparisons");
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Returns time series configuration for the given group of checks.
     *
     * @param tableSpec Parent table specification - used to get the details about the time partitioning column.
     * @return Time series configuration.
     */
    @Override
    public TimeSeriesConfigurationSpec getTimeSeriesConfiguration(TableSpec tableSpec) {
        return new TimeSeriesConfigurationSpec()
        {{
            setMode(TimeSeriesMode.current_time);
            setTimeGradient(TimePeriodGradient.month);
        }};
    }

    /**
     * Returns the type of checks (profiling, recurring, partitioned).
     *
     * @return Check type.
     */
    @Override
    @JsonIgnore
    public CheckType getCheckType() {
        return CheckType.recurring;
    }

    /**
     * Returns the time range for recurring and partitioned checks (daily, monthly, etc.).
     * Profiling checks do not have a time range and return null.
     *
     * @return Time range (daily, monthly, ...).
     */
    @Override
    @JsonIgnore
    public CheckTimeScale getCheckTimeScale() {
        return CheckTimeScale.monthly;
    }

    /**
     * Returns the check target, where the check could be applied.
     *
     * @return Check target, "table" or "column".
     */
    @Override
    @JsonIgnore
    public CheckTarget getCheckTarget() {
        return CheckTarget.table;
    }

    /**
     * Returns the name of the cron expression that is used to schedule checks in this check root object.
     *
     * @return Recurring schedule group (named schedule) that is used to schedule the checks in this root.
     */
    @Override
    @JsonIgnore
    public CheckRunRecurringScheduleGroup getSchedulingGroup() {
        return CheckRunRecurringScheduleGroup.recurring_monthly;
    }
}
